File: 백준/Gold/14502. 연구소/연구소.py
from collections import deque
import sys
import copy
input = sys.stdin.readline

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

def makeWall(count):
  if count == 3:
    bfs()
    return
  
  for i in range(n):
    for j in range(m):
      if lab[i][j] == 0:
        lab[i][j] = 1
        makeWall(count + 1)
        lab[i][j] = 0

def bfs():
  global ans
  safe = 0
  queue = deque()
  visited = [[False for _ in range(m)] for _ in range(n)]

  tmp_arr = copy.deepcopy(lab)

  for i in range(n):
    for j in range(m):
      if tmp_arr[i][j] == 2:
        queue.append((i,j))
        visited[i][j] = True

  while queue:
    x, y = queue.popleft()

    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]

      if nx < 0 or ny < 0 or nx >= n or ny >= m:
        continue
      if visited[nx][ny] == False and tmp_arr[nx][ny] == 0:
        queue.append((nx, ny))
        visited[nx][ny] = True
        tmp_arr[nx][ny] = 2

  for i in range(n):
    safe += tmp_arr[i].count(0)
  
  ans = max(ans, safe)

n, m = map(int, input().split())
lab = [list(map(int, input().split())) for _ in range(n)]
ans = 0

makeWall(0)

print(ans)

File: 백준/Gold/14502. 연구소/연구소.java
import java.util.*;
import java.io.*;

public class Main {

    static int N, M;
    static int[][] lab = new int[9][9];

    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, 1, 0, -1};

    static int answer;

    static class Pos {
        int x;
        int y;

        public Pos(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0; j < M; j++) {
                lab[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        buildWall(0);

        System.out.println(answer);
    }

    private static void buildWall(int count) {
        if (count == 3) {
            bfs();
            return;
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (lab[i][j] == 0) {
                    lab[i][j] = 1;
                    buildWall(count + 1);
                    lab[i][j] = 0;
                }
            }
        }

    }

    private static void bfs() {
        int[][] tempLab = new int[N][M];
        boolean[][] visited = new boolean[N][M];

        Deque<Pos> q = new ArrayDeque<>();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                tempLab[i][j] = lab[i][j];

                if (tempLab[i][j] == 2) {
                    q.addLast(new Pos(i, j));
                    visited[i][j] = true;
                }

            }
        }

        while (!q.isEmpty()) {
            Pos curPos = q.removeFirst();

            for (int i = 0; i < 4; i++) {
                int nx = curPos.x + dx[i];
                int ny = curPos.y + dy[i];

                if (nx < 0 || ny < 0 || nx >= N || ny >= M) {
                    continue;
                }

                if (!visited[nx][ny] && tempLab[nx][ny] == 0) {
                    q.addLast(new Pos(nx, ny));
                    visited[nx][ny] = true;
                    tempLab[nx][ny] = 2;
                }

            }

        }

        int safeArea = countSafe(tempLab);

        answer = Math.max(safeArea, answer);

    }

    private static int countSafe(int[][] map) {

        int safeArea = 0;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (map[i][j] == 0) {
                    safeArea += 1;
                }

            }
        }

        return safeArea;
    }
}
